src/zarr/storage/_common.py

```diff
@@ -54,59 +62,87 @@ def __init__(self, store: Store, path: str = "") -> None:
         """
         Open StorePath based on the provided mode.
 
-        * If the mode is 'w-' and the StorePath contains keys, raise a FileExistsError.
-        * If the mode is 'w', delete all keys nested within the StorePath
-        * If the mode is 'a', 'r', or 'r+', do nothing
+        * If the mode is None, return an opened version of the store with no changes.
+        * If the mode is 'r+', 'w-', 'w', or 'a' and the store is read-only, raise a ValueError.
+        * If the mode is 'r' and the store is not read-only, return a copy of the store with read_only set to True.
+        * If the mode is 'w-' and the store is not read-only and the StorePath contains keys, raise a FileExistsError.
+        * If the mode is 'w'  and the store is not read-only, delete all keys nested within the StorePath.
 
         Parameters
         ----------
         mode : AccessModeLiteral
             The mode to use when initializing the store path.
 
+            The accepted values are:
+
+            - ``'r'``: read only (must exist)
+            - ``'r+'``: read/write (must exist)
+            - ``'a'``: read/write (create if doesn't exist)
+            - ``'w'``: read/write (overwrite if exists)
+            - ``'w-'``: read/write (create if doesn't exist).
+
         Raises
         ------
         FileExistsError
             If the mode is 'w-' and the store path already exists.
         ValueError
             If the mode is not "r" and the store is read-only, or
-            if the mode is "r" and the store is not read-only.
         """
 
-        await store._ensure_open()
-        self = cls(store, path)
-
         # fastpath if mode is None
         if mode is None:
-            return self
+            return await cls._create_open_instance(store, path)
 
-        if store.read_only and mode != "r":
-            raise ValueError(f"Store is read-only but mode is '{mode}'")
-        if not store.read_only and mode == "r":
-            raise ValueError(f"Store is not read-only but mode is '{mode}'")
+        if mode not in ANY_ACCESS_MODE:
+            raise ValueError(f"Invalid mode: {mode}, expected one of {ANY_ACCESS_MODE}")
 
+        if store.read_only:
+            # Don't allow write operations on a read-only store
+            if mode != "r":
+                raise ValueError(
+                    f"Store is read-only but mode is {mode!r}. Create a writable store or use 'r' mode."
+                )
+            self = await cls._create_open_instance(store, path)
+        elif mode == "r":
+            # Create read-only copy for read mode on writable store
+            try:
+                warnings.warn(
+                    "Store is not read-only but mode is 'r'. Creating a read-only copy. "
+                    "This behavior may change in the future with a more granular permissions model.",
+                    UserWarning,
+                    stacklevel=1,
+                )
```

Comment:
I feel like this warning could be outside of the `try... except` block. The start of the message is redundant with that in the re-raised `ValueError`, so why not just move this warning to before the `try...except`, effectively factoring out that part of the message?

Reply:
Motivated by the idea that it's good practice to have the `try...except` wrap the absolute minimum amount of code that could raise to minimize the chance of it raising on something unexpected.

Reply:
In fact going even further I guess the `try... except` only needs to wrap 

```python
read_only_store = store.with_read_only(True)
```
not the outer part
```python
self = await cls._create_open_instance(read_only_store, path)
```

Reply:
thanks for the suggestions, I removed the warning in [`4e099e4` (#3156)](https://github.com/zarr-developers/zarr-python/pull/3156/commits/4e099e4f8c49090b182cf1c60e25006446f26253) and shortened the try...except in [`2616b2f` (#3156)](https://github.com/zarr-developers/zarr-python/pull/3156/commits/2616b2f8d030e6adfede6202c88d5331b3992446)